### PR TITLE
Make VSTest runner work and enable html log

### PR DESCRIPTION
**Pull request by ViktorHofer**
The VSTest runner is now supposed to work with the `/p:TestRunner=vstest`
switch. Currently this is an opt-in but the desire is to get rid of
xunit.console completely and just use VSTest.

An html and trx log example for System.Text.RegularExpressions.Tests: [testResults.zip](https://github.com/dotnet/runtime/files/4362888/testResults.zip)

cc @stephentoub @danmosemsft

### CODE REVIEW

**Comment by stephentoub** [comment]
Thanks. What does an error like this mean?
```
Non-serializable data ('System.Object[]') found for 'System.Text.RegularExpressions.Tests.RegexConstructorTests.Ctor'; falling back to single test case.
```
Did the test run and it just can't report on it, or did it skip it?

**Comment by ViktorHofer** [comment]
https://github.com/dotnet/runtime/issues/30324

> Did the test run and it just can't report on it, or did it skip it?

It ran but it can't fetch individual data rows, means you won't be able to debug individual data rows in VS Test Explorer. I can only hide those diagnostic messages by disabling all, which might be possible as we only used them for the long running test support for which VSTest might have integrated support.

**Comment by jkotas** [comment]
Are we going to have a workflow in place to make changes in the VSTest runner that are required for not-yet-released .NET Core versions?

For example, let's say that that we add a new platform and [this](https://github.com/microsoft/vstest/blob/master/src/Microsoft.TestPlatform.PlatformAbstractions/netcore/System/PlatformEnvironment.cs) starts failing. How we are going to make progress?

I am asking about this because of I am stuck on a similar problem with BenchmarkDotNet in 33694.

**Comment by ViktorHofer** [comment]
We already consume vstest via dependency flow which means getting live updates which aren't bound to VS releases. Additionally to that, we are actively collaborating with the vstest team (cc @nohwnd) and benefiting from each other.

File: eng/testing/xunit/vstest.props [hunk 1]
```diff
@@ -2,19 +2,15 @@
   <PropertyGroup>
     <!-- Microsoft.Net.Test.Sdk brings a lot of satellite assemblies in. -->
     <SatelliteResourceLanguages>en</SatelliteResourceLanguages>
+    <TrxTestResultsName>testResults.trx</TrxTestResultsName>
```
Comment (safern) [inline eng/testing/xunit/vstest.props]: Do we know if the helix sdk supports .trx files to show the test results in AzDo, submitting to kusto and failing the build? If not we should request a feature for when we want to move to use this as our default test runner. cc: @alexperovich

File: eng/testing/xunit/vstest.targets [hunk 2]
```diff
@@ -1,10 +1,13 @@
 <Project>
   <PropertyGroup>
-    <RunCommand>"$(RunScriptHost)"</RunCommand>
+    <RunCommand>"$(DotNetTool)"</RunCommand>
+    <RunCommand Condition="'$(ContinuousIntegrationBuild)' == 'true'">dotnet</RunCommand>
```
Comment (safern) [inline eng/testing/xunit/vstest.targets]: Just to understand, could you explain what's the difference in between `DotNetTool` and `dotnet` and why does it has to be different in CI?

File: eng/testing/xunit/vstest.props (hunk 1 above)
Comment (alexperovich) [reply eng/testing/xunit/vstest.props]: Helix supports trx files for uploading to azure devops. Kusto is only supported for xunit.

File: eng/testing/xunit/vstest.xunit.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+    "diagnosticMessages": false,
+    "longRunningTestSeconds": 120,
```
Comment (safern) [inline eng/testing/xunit/vstest.xunit.json]: So this support is now built-in in VSTest?

File: eng/Versions.props [hunk 4]
```diff
@@ -118,7 +118,6 @@
     <XUnitVersion>2.4.1</XUnitVersion>
     <TraceEventVersion>2.0.5</TraceEventVersion>
     <NewtonsoftJsonVersion>12.0.3</NewtonsoftJsonVersion>
-    <XUnitXmlTestLoggerVersion>2.1.26</XUnitXmlTestLoggerVersion>
```
Comment (safern) [inline eng/Versions.props]: Why is this removed? Just left-over?

File: eng/testing/xunit/vstest.props (hunk 1 above)
Comment (safern) [reply eng/testing/xunit/vstest.props]: I think a lot of people use Kusto for metrics and what not as the AzDo APIs aren't that great to get test results and search through them. Would it be possible to add Kusto from a trx file?

Comment (alexperovich) [reply eng/testing/xunit/vstest.props]: The support for processing xunit results into kusto is on the server side for helix and extending that for other formats would be a very large an potentially disruptive change. The azure devops uploading code is in arcade and very easy to change so it supports many formats. My understanding is that we want to consolidate everything to azure devops and make that experience better, rather than continue investing in our custom reporting. @Chrisboh may have more context.

Comment (ViktorHofer) [reply eng/testing/xunit/vstest.props]: Yeah I knew that trx is supported but I didn't know about Kusto only working with the xunit xml format.

File: eng/testing/xunit/vstest.targets (hunk 2 above)
Comment (ViktorHofer) [reply eng/testing/xunit/vstest.targets]: DotNetTool is a property set by Arcade, mean it's available during the test invocation on your machine. On a different machine you don't want to hardcode the path into the RunTemplate.cmd/sh. There is more work required here anyway to make this work on CI.

My goal in a subsequent PR is to just use dotnet here instead and then for local local development, modify PATH to point to the RunScriptHost if not on CI. As said, I don't want to do this now as this just enables VSTest optionally and not on CI.

Comment (ViktorHofer) [reply eng/testing/xunit/vstest.targets]: Another blocker for enabling VSTest in CI is https://github.com/dotnet/runtime/issues/1980.

File: eng/testing/xunit/vstest.xunit.json (hunk 3 above)
Comment (ViktorHofer) [reply eng/testing/xunit/vstest.xunit.json]: Yeah but I still need to play around with it. There's a blame mode which creates a sequence.xml which tells you which tests have run and which haven't but I don't know if that mode slows down the test invocation. cc @nohwnd

File: eng/Versions.props (hunk 4 above)
Comment (ViktorHofer) [reply eng/Versions.props]: Yeah this was introduced only for vstest in a past PR but actually we don't need it.

**Review by safern — APPROVED** [review]
Thanks for solving all my questions.

**Comment by ViktorHofer** [comment]
Issue **was** https://github.com/dotnet/runtime/issues/32377.
